doc/dev/adr/0025-rpc-streaming.md

```diff
@@ -0,0 +1,400 @@
+- [Bi-directional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) (many request messages, many response messages. Request and response stream may send concurrently and/or in any order)
+
+[gRPC also allows for either the client or server to cancel an RPC at any time](https://grpc.io/docs/what-is-grpc/core-concepts/#cancelling-an-rpc)
+
+## Decision
+
+### API design, .NET
+
+While RPC streaming shares a lot of similarities to normal RPC, we will define a new communication pattern to handle this scenario with two corresponding base classes: ```StreamingCommandInvoker``` and ```StreamingCommandExecutor```. 
+
+These new base classes will use similar versions of the ```ExtendedRequest``` and ```ExtendedResponse``` RPC classes to include the streaming-specific information about each request and response:
+
+```csharp
+public class StreamingExtendedRequest<TReq>
+    where TReq : class
+{
+    /// <summary>
+    /// The request payload
+    /// </summary>
+    public TReq Request { get; set; }
+
+    /// <summary>
+    /// The metadata specific to this message in the stream
+    /// </summary>
+    public StreamMessageMetadata Metadata { get; set; }
+}
+
+public class StreamingExtendedResponse<TResp>
+        where TResp : class
+{
+    /// <summary>
+    /// The response payload
+    /// </summary>
+    public TResp Response { get; set; }
+
+    /// <summary>
+    /// The metadata specific to this message in the stream
+    /// </summary>
+    public StreamMessageMetadata Metadata { get; set; }
+}
+```
+
+#### Invoker side
+
+The new API will ask users to provide a stream of request payloads + metadata, the target streaming command executor, any stream-level metadata and timeout/cancellation tokens. It will return the
+stream of responses as well as a cancellation function that allows the user to terminate the stream exchange at any time. 
+
+```csharp
+public abstract class StreamingCommandInvoker<TReq, TResp>
+    where TReq : class
+    where TResp : class
+{
+    /// <summary>
+    /// Invoke a streaming command on a particular streaming command executor
+    /// </summary>
+    /// <param name="requests">The stream of requests to send. This stream must contain at least one request.</param>
+    /// <param name="streamMetadata">The metadata for the request stream as a whole.</param>
+    /// <param name="additionalTopicTokenMap">Topic tokens to substitute in the request topic.</param>
+    /// <param name="streamExchangeTimeout">The timeout between the beginning of the request stream and the end of both the request and response stream.</param>
+    /// <param name="cancellationToken">Cancellation token. Signalling this will also make a single attempt to notify the executor of the cancellation.</param>
+    /// <returns>The stream of responses.</returns>
+    public async Task<IStreamContext<StreamingExtendedResponse<TResp?>> InvokeStreamingCommandAsync(
```

Comment:
Can we please use `IAsyncEnumerable<StreamingExtendedResponse<TResp?>>` as return value?